### PR TITLE
[ci] Reduce runners load for pulls that affect only doc or cpp changes

### PR DESCRIPTION
**Pull request by nicoloboschi**
### Motivation
1. For doc or cpp changes now the Pulsar CI workflow requires runners to just skip steps.
2. CPP builds run on different workflows and this can be optimized in a single workflow
3. MacOS build runner is created even if there are only doc changes
4. GO functions check can reuse the same runner

### Modifications

* Moved the check for doc changes to jobs level instead of steps. In case the condition of `if` returns false, no new runners will be spawned
* Moved all CPP builds to the same workflow
* Moved go functions checks to the same runner
* Moved MacOs build as step after Pulsar CI workflow

 
- [x] `doc-not-needed`

### CODE REVIEW

**Comment by lhotari** [comment]
> Moved the check for doc changes to jobs level instead of steps. In case the condition of if returns false, no new runners will be spawned

That solution is possible only when the check produced by the skipped build job isn't required That's why the skipping solution is needed at the build step level. It shouldn't be done at such a fine grain level. A workflow where the last step is the required check would be a way to solve this so that the builds wouldn't have to be scheduled at all. This would also simplify the workflows since we could remove the step level skipping.

**Review by lhotari — CHANGES_REQUESTED** [review]
Please modify the workflows in a way so that you add a new fan-in build job called "{workflow name} - Checks completed" from all build jobs that previously used to be listed in .asf.yaml in the required checks. When we re-add required checks to .asf.yaml, we would only add the  "{workflow name} - Checks completed" check as required. This way we can do the nice optimization that you have performed in this PR.

**Comment by nicoloboschi** [comment]
I moved also the non required checks to the flaky workflow and now we have the latest jobs for Pulsar CI that will need to be put as required for merging

File: .github/workflows/pulsar-ci.yaml [hunk 1]
```diff
@@ -598,9 +593,6 @@ jobs:
           - name: Schema
             group: SCHEMA
 
-          - name: Plugin
-            group: PLUGIN
-
```
Comment (lhotari) [inline .github/workflows/pulsar-ci.yaml:603]: Is this intentionally removed?

File: .github/workflows/pulsar-ci.yaml [hunk 2]
```diff
@@ -762,4 +858,4 @@ jobs:
 
       - name: Delete maven repository binaries from GitHub Actions Artifacts
         run: |
```
Comment (lhotari) [inline .github/workflows/pulsar-ci.yaml:860]: `if: ${{ needs.changed_files_job.outputs.docs_only != 'true' }}` is needed here.

File: .github/workflows/pulsar-ci.yaml [hunk 3]
```diff
@@ -610,30 +602,135 @@ jobs:
...
+
+      - name: Wait for ssh connection when build fails
+        # ssh access is enabled for builds in own forks
+        uses: ./.github/actions/ssh-access
+        if: ${{ failure() && github.repository != 'apache/pulsar' }}
+        continue-on-error: true
+        with:
+          action: wait
+
+  flaky-system-tests:
```
Comment (lhotari) [inline .github/workflows/pulsar-ci.yaml:700]: what is the exact reason for adding this? I would assume that it's possible to add matrix variables to cover this.

Comment (nicoloboschi) [reply .github/workflows/pulsar-ci.yaml:700]: yes but then the new step `pulsar-ci-checks-completed` needs these tests passing. Currently in the `.asf.yaml` they are not listed. So now it depends only on `system-tests` job and the result of this one doesn't impact the mergeablity of the pull

Comment (lhotari) [reply .github/workflows/pulsar-ci.yaml:700]: oh yes you are right.

**Review by lhotari — APPROVED** [review]

**Pull request merged**